diff -urN a/micctrl/micctrl.c b/micctrl/micctrl.c
--- a/micctrl/micctrl.c	2014-01-07 19:30:33.000000000 +0100
+++ b/micctrl/micctrl.c	2014-01-07 20:07:59.000000000 +0100
@@ -1146,7 +1146,7 @@
 	int lockfd;
 	int err;
 
-	if ((lockfd = open(LOCKFILE_NAME, O_WRONLY | O_CREAT)) < 0) {
+	if ((lockfd = open(LOCKFILE_NAME, O_WRONLY | O_CREAT, 0644)) < 0) {
 		fprintf(stderr, "Could not open the lockfile '%s\n", LOCKFILE_NAME);
 		exit(1);
 	}
@@ -2463,7 +2463,7 @@
 		flen = read(fd, buf, sbuf.st_size);
 		close(fd);
 
-		if ((fd = open(oldto, O_WRONLY | O_CREAT), sbuf.st_mode & 0777) < 0) {
+		if ((fd = open(oldto, O_WRONLY | O_CREAT, sbuf.st_mode & 0777)) < 0) {
 			free(buf);
 			return;
 		}
diff -urN a/micctrl/network.c b/micctrl/network.c
--- a/micctrl/network.c	2014-01-07 19:30:34.000000000 +0100
+++ b/micctrl/network.c	2014-01-07 20:50:45.000000000 +0100
@@ -648,7 +648,7 @@
 	rlen = read(fd, buffer, sbuf.st_size);
 	close(fd);
 
-	if ((fd = open(oname, O_WRONLY|O_CREAT)) < 0) {
+	if ((fd = open(oname, O_WRONLY|O_CREAT, 0644)) < 0) {
 		display(DERROR, "%s: Could not open %s\n", mic->name, oname);
 		free(buffer);
 		return 0;
@@ -661,7 +661,7 @@
 	    (mic->config.rootdev.type == ROOT_TYPE_SPLITNFS)) {
 		mpss_filename(oname, PATH_MAX, "%s/etc/ssh/%s",
 			      strchr(mic->config.rootdev.target, ':') + 1, name);
-		if ((fd = open(oname, O_WRONLY|O_CREAT)) < 0) {
+		if ((fd = open(oname, O_WRONLY|O_CREAT, 0644)) < 0) {
 			display(DERROR, "%s: Could not open %s\n", mic->name, oname);
 			free(buffer);
 			return 0;
@@ -3770,7 +3770,7 @@
 	mac[0] &= 0xfe;	// clear multicast bit
 	mac[0] |= 0x02;	// set local assignment bit
 
-	snprintf(buf, 128, "%02x:%02x:%02x:%02x:%02x:%02x",
+	snprintf(buf, 32, "%02x:%02x:%02x:%02x:%02x:%02x",
 			   mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
 
 	if ((tmpaddr = malloc(strlen(buf) + 1)) != NULL) {
diff -urN a/mpssd/mpssd.c b/mpssd/mpssd.c
--- a/mpssd/mpssd.c	2014-01-07 19:30:34.000000000 +0100
+++ b/mpssd/mpssd.c	2014-01-07 19:58:47.000000000 +0100
@@ -516,7 +516,7 @@
 	struct flock fl;
 	int sc;
 
-	if ((lockfd = open(LOCKFILE_NAME, O_WRONLY | O_CREAT)) < 0) {
+	if ((lockfd = open(LOCKFILE_NAME, O_WRONLY | O_CREAT, 0644)) < 0) {
 		fprintf(stderr, "cannot open lockfile '%s'\n", LOCKFILE_NAME);
 		exit(EBADF);
 	}
@@ -1241,7 +1241,7 @@
 		len = read(fd, cookie, sz);
 		close(fd);
 
-		if ((fd = open(filename, O_WRONLY|O_CREAT)) < 0) {
+		if ((fd = open(filename, O_WRONLY|O_CREAT, 0644)) < 0) {
 			mpsslog(DERROR, "Failed to open %s\n", strerror(errno));
 			return;
 		}
